### PR TITLE
Cache the active realm information in Identities

### LINKED ISSUE #4276: Cache the active realm information in Identities
**Motivation**
Identities allows to verify the token of the caller.
One of the steps is checking that the caller belongs to an active realm.
Currently a call to the database is done each time, we could avoid that as:
* A realm is almost never deprecated
* We already rely on cached information for the user groups

**Acceptance criteria**
1. A second LRU cache with the same configuration as the groups one is introduced for realms

### PR DESCRIPTION
**Pull request by olivergrabinski**
Closes #4276

### CODE REVIEW

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/identities/IdentitiesImpl.scala [hunk 1]
```diff
@@ -61,6 +62,14 @@ class IdentitiesImpl private[identities] (
       )
     }
 
+    def fetchRealm(parsedToken: ParsedToken): IO[Realm] =
+      realm
+        .getOrElseUpdate(parsedToken.rawToken, findActiveRealm(parsedToken.issuer))
+        .flatMap {
+          case Some(realm) => IO.pure(realm)
+          case None        => IO.raiseError(UnknownAccessTokenIssuer)
+        }
```
Comment (dantb) [inline delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/identities/IdentitiesImpl.scala:71]: Stylistic nit (not sure if we use monad transformers but they can be useful tools for isolated things like this)
```suggestion
      OptionT(realm.getOrElseUpdate(parsedToken.rawToken, findActiveRealm(parsedToken.issuer))).getOrRaise(UnknownAccessTokenIssuer)
```

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/identities/IdentitiesImpl.scala [hunk 2]
```diff
@@ -133,10 +142,14 @@ object IdentitiesImpl {
     *   the cache configuration
     */
   def apply(realms: Realms, hc: HttpClient, config: CacheConfig): IO[Identities] = {
+    val groupsCache = LocalCache[String, Set[Group]](config)
+    val realmCache  = LocalCache[String, Option[Realm]](config)
```
Comment (dantb) [inline delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/identities/IdentitiesImpl.scala:146]: Not necessarily for this change but [mules](https://github.com/davenverse/mules/) is a nice alternative to caffeine cache (cats effect friendly and optimised)

**Review by dantb — DISMISSED** [review]
LGTM

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/identities/IdentitiesImpl.scala [hunk 3]
```diff
@@ -61,6 +62,14 @@ class IdentitiesImpl private[identities] (
       )
     }
 
+    def fetchRealm(parsedToken: ParsedToken): IO[Realm] =
+      realm
+        .getOrElseUpdate(parsedToken.rawToken, findActiveRealm(parsedToken.issuer))
```
Comment (imsdu) [inline delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/identities/IdentitiesImpl.scala:67]: There is `getOrElseAttemptUpdate` that may be a better fit

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/identities/IdentitiesImpl.scala [hunk 4]
```diff
@@ -61,6 +62,14 @@ class IdentitiesImpl private[identities] (
...
+        .flatMap {
+          case Some(realm) => IO.pure(realm)
```
Comment (imsdu) [inline delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/identities/IdentitiesImpl.scala:69]: `IO.fromOption` / `IO.raiseWhen` ?

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/identities/IdentitiesImpl.scala (hunk 2 above)
Comment (imsdu) [reply delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/identities/IdentitiesImpl.scala:146]: There is a caffeine dependency so it looks like it is using it behind the scene

Comment (dantb) [reply delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/identities/IdentitiesImpl.scala:146]: Oh yeah, I think that must be if you just want a functional wrapper but no change in behaviour (maybe that is preferable for us?). [`MemoryCache`](https://github.com/davenverse/mules/blob/main/modules/core/src/main/scala/io/chrisdavenport/mules/MemoryCache.scala#L311-L323) does the whole thing using CE (`MapRef` in the implementation)

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/identities/IdentitiesImpl.scala (hunk 1 above)
Comment (olivergrabinski) [reply delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/identities/IdentitiesImpl.scala]: I did a variation of that

File: delta/sdk/src/test/scala/ch/epfl/bluebrain/nexus/delta/sdk/identities/IdentitiesImplSuite.scala [hunk 5]
```diff
@@ -326,4 +330,26 @@ class IdentitiesImplSuite extends CatsEffectSuite with TestHelpers with IOFromMa
...
+      useCommas = true
+    )
+
+    for {
+      parsedToken <- IO.fromEither(ParsedToken.fromToken(token))
+      realm       <- realmCache
+      groups      <- groupsCache
+      _           <- realm.get(parsedToken.rawToken).assertNone
+      _           <- groups.get(parsedToken.rawToken).assertNone
+      _           <- identitiesFromCaches(realm, groups).exchange(token)
```
[the remark this answers is not in the document]
Comment (olivergrabinski) [reply delta/sdk/src/test/scala/ch/epfl/bluebrain/nexus/delta/sdk/identities/IdentitiesImplSuite.scala]: I added at test to cover it now, maybe you know a better way of doing it

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/identities/IdentitiesImpl.scala (hunk 3 above)
Comment (olivergrabinski) [reply delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/identities/IdentitiesImpl.scala]: Done

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/identities/IdentitiesImpl.scala (hunk 4 above)
Comment (olivergrabinski) [reply delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/identities/IdentitiesImpl.scala]: Ended up using OptionT, what do you think?

**Pull request merged**